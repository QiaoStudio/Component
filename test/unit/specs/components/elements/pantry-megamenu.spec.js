import {
  shallowMount
} from 'dp-test'
import ComMegamenu from 'components/elements/megamenu/megamenu.vue'
import sinon from 'sinon'
import _ from 'lodash'

describe('elements/Megamenu', () => {
  let wrapper
  let sanbox
  const columns = [{
    'title': 'Compare insurance',
    'type': 'link',
    'layout': 'two-column',
    'items': [{
      'iconType': 'product',
      'iconCategory': 'travel-insurance',
      'iconDisplay': 'dp-icon--sm-only',
      'label': 'Travel Insurance',
      'url': '/sg/travel-insurance',
      'target': ''
    },
    {
      'iconType': 'product',
      'iconCategory': 'car-insurance',
      'iconDisplay': 'dp-icon--sm-only',
      'label': 'Car Insurance',
      'url': '/sg/car-insurance',
      'target': ''
    },
    {
      'iconType': 'product',
      'iconCategory': 'health-insurance',
      'iconDisplay': 'dp-icon--sm-only',
      'label': 'Driver & Passenger\' <br>Personal Accident',
      'url': '/sg/personal-accident',
      'target': ''
    }]
  },
  {
    'title': 'Promotion',
    'type': 'article',
    'layout': '',
    'items': [{
      'imgUrl': 'https://picsum.photos/120',
      'url': '/sg/promotion',
      'title': 'Learn a bunch of stuff here',
      'description': 'Don\'t miss the exclusive campaign because you don\'t want to.',
      'target': '_self'
    }]
  }]

  function createWrapper(options) {
    let shallowOption = _.merge({
      data() {
        return {
          columns: []
        }
      }
    }, options || {})
    return shallowMount(ComMegamenu, shallowOption)
  }

  beforeEach(() => {
    sanbox = sinon.createSandbox()
    wrapper = createWrapper()
  })

  afterEach(() => {
    wrapper.destroy()
    sanbox.restore()
  })

  it('when megamenu has columns, megamenu should match snapshot', () => {
    wrapper = createWrapper({
      propsData: {
        columns: columns
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('when megamenu has columns, megamenu columns should be array', () => {
    wrapper = createWrapper({
      propsData: {
        columns: columns
      }
    })
    expect(Array.isArray(wrapper.vm.columns)).toBe(true)
  })

  it('when megamenu has links, megamenu links should be array', () => {
    wrapper = createWrapper({
      propsData: {
        columns: columns
      }
    })
    expect(Array.isArray(wrapper.vm.columns[0].items)).toBe(true)
  })

  it('when megamenu has article, megamenu article should be array', () => {
    wrapper = createWrapper({
      propsData: {
        columns: columns
      }
    })
    expect(Array.isArray(wrapper.vm.columns[1].items)).toBe(true)
  })

  it('when megamenu has two columns, megamenu two columns should be shown', () => {
    wrapper = createWrapper({
      propsData: {
        columns: columns
      }
    })
    expect(wrapper.findAll('.dp-megamenu > div')).toHaveLength(2)
  })

  it('when megamenu links are set, megamenu links should be shown', () => {
    wrapper = createWrapper({
      propsData: {
        columns: columns
      }
    })
    expect(wrapper.findAll('.dp-megamenu__link')).toHaveLength(3)
  })

  it('when megamenu article is set, megamenu article should be shown', () => {
    wrapper = createWrapper({
      propsData: {
        columns: columns
      }
    })
    expect(wrapper.findAll('.dp-megamenu__article')).toHaveLength(1)
  })
})
