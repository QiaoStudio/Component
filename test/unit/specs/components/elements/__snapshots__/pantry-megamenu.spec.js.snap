// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elements/Megamenu when megamenu has columns, megamenu should match snapshot 1`] = `
<div class="dp-megamenu">
  <div><span>Compare insurance</span>
    <ul class="dp-megamenu__link--two-column">
      <li class="dp-megamenu__link">
        <a href="/sg/travel-insurance" target="_self">
          <dp-icon-stub type="product" category="travel-insurance" display="dp-icon--sm-only"></dp-icon-stub> <span class="dp-megamenu__link-label">Travel Insurance</span></a>
      </li>
      <li class="dp-megamenu__link">
        <a href="/sg/car-insurance" target="_self">
          <dp-icon-stub type="product" category="car-insurance" display="dp-icon--sm-only"></dp-icon-stub> <span class="dp-megamenu__link-label">Car Insurance</span></a>
      </li>
      <li class="dp-megamenu__link">
        <a href="/sg/personal-accident" target="_self">
          <dp-icon-stub type="product" category="health-insurance" display="dp-icon--sm-only"></dp-icon-stub> <span class="dp-megamenu__link-label">Driver &amp; Passenger' <br>Personal Accident</span></a>
      </li>
    </ul>
  </div>
  <div><span>Promotion</span>
    <div>
      <div class="dp-megamenu__article">
        <a href="/sg/promotion" target="_self">
          <div class="dp-megamenu__article-img">
            <img src="https://picsum.photos/120" alt="Learn a bunch of stuff here">
          </div>
          <div><span class="dp-megamenu__article-title">Learn a bunch of stuff here</span>
            <p class="dp-megamenu__article-desc">Don't miss the exclusive campaign because you don't want to.</p>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
